UnitTests/GitUITests/UITest.cs

```diff
@@ -27,29 +29,41 @@ void HandleApplicationIdle(object sender, EventArgs e)
         }
 
         public static void RunForm<T>(
-            Action showDialog,
+            Action showForm,
             Func<T, Task> runAsync)
             where T : Form
         {
+            // Start runAsync before calling showForm.
+            // The latter might block until the form is closed, especially if using Application.Run(form).
+
             // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
             // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
             // complete.
             var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
+                // Wait for the form to be opened by the test thread.
+                T form;
+                do
+                {
+                    await WaitForIdleAsync(); // does also SwitchToMainThreadAsync
+                    form = Application.OpenForms.OfType<T>().SingleOrDefault();
+                }
+                while (form == null || !form.Visible);
+
+                // Wait for potential pending asynchronous tasks triggered by the form.
+                ThreadHelper.WaitForPendingOperationsAndDoApplicationEvents();
```

Comment:
❓ In what specific case can this be required?

📝 The contract of `WaitForIdleAsync` should guarantee that this call is not required, so I would need to see a detailed sequence explaining why this change is necessary.

Reply:
It's not about `WaitForIdleAsync` but `WaitForPendingOperations`.
Does it really guarentee that Windows messages are processed?
(The inline documentation of `JoinTillEmptyAsync` is rather short.)
I am quite sure that I've seen several test runs which hung in the creation of a form.

Reply:
> Does it really guarentee that Windows messages are processed?

No. But surrounding calls of `Application.DoEvent` before and after `AsyncTestHelper.WaitForPendingOperations(AsyncTestHelper.UnexpectedTimeout)` suffice.